src/controllers/media.ts

```diff
@@ -338,3 +338,147 @@ export const getByImdbID = async(ctx: ParameterizedContext<any, Router.IRouterPa
     throw new MediaNotFoundError();
   }
 };
+
+export const getVideo = async(ctx: ParameterizedContext<any, Router.IRouterParamContext<any, {}>>): Promise<any> => {
+  const { title, osdbHash, imdbID } = ctx.query;
+  let { episode, season, year, filebytesize } = ctx.query;
+  [episode, season, year, filebytesize] = [episode, season, year, filebytesize].map(param => param ? Number(param) : null);
+
+  if (!title && !osdbHash && !imdbID) {
+    throw new ValidationError('title, osdbHash or imdbId is a required parameter');
+  }
+
+  if (osdbHash && !filebytesize) {
+    throw new ValidationError('filebytesize is required when passing osdbHash');
+  }
+  let isOnlyOpenSubsSearch = false;
+
+  const query = [];
+  const failedQuery = [];
+
+  if (osdbHash) {
+    query.push({ osdbHash });
+    failedQuery.push({ osdbHash });
+    if (!imdbID && !title) {
+      isOnlyOpenSubsSearch = true;
+    }
+  }
+
+  if (imdbID) {
+    query.push({ imdbID });
+    failedQuery.push({ imdbId: imdbID });
+  }
+
+  if (title) {
+    const titleQuery: any = { searchMatches: { $in: [title] } };
+    const titleFailedQuery: any = { title };
+
+    if (year) {
+      titleQuery.year = year;
+      titleFailedQuery.year = year;
+    }
+    if (episode) {
+      titleQuery.episode = episode;
+      titleFailedQuery.episode = episode;
+    }
+    if (season) {
+      titleQuery.season = season;
+      titleFailedQuery.season = season;
+    }
+    query.push(titleQuery);
+    failedQuery.push(titleFailedQuery);
+  }
+
+  // find an existing metadata record, or previous failure record
+  const [existingResult, existingFailedResult] = await Promise.all([
+    MediaMetadata.findOne({ $or: query }, null, { lean: true }).exec(),
+    FailedLookups.findOne({ $or: failedQuery }, null, { lean: true }).exec(),
+  ]);
+
+  // we have an existing metadata record, so return it
+  if (existingResult) {
+    return ctx.body = existingResult;
+  }
+
+  // we have an existing failure record, so increment it, and throw not found error
+  if (existingFailedResult) {
+    await FailedLookups.updateOne({ _id: existingFailedResult._id }, { $inc: { count: 1 } }).exec();
+    throw new MediaNotFoundError();
+  }
+
+  // the database does not have a record of this file, so begin search for metadata on external apis.
+
+  // Start OpenSubtitles lookups
```

Comment:
I had a thought that we could do the OMDb lookup before the OSDb one (and I made a client-side change to do that already) since that's the most likely to succeed. If you agree could you swap the order here?

Reply:
We get better results for osdbHash only results doing this first, as we can then use the found imdbId to call OMDb to get richer data, but we lose that if we do this first

Reply:
I was getting some really long load times for OSDb lookups the other day, seemed like there was some 30 second timeout or something at play. But yeah ok.